### DISCUSSION TITLE
Parameter type annotations for `Call service` automation UI? (fancy parameter selection)

### DISCUSSION THREAD

**Discussion by HealsCodes** [custom-components/pyscript#329] [Q&A] [answered]
Hi,
I hope this wasn't already answered before otherwise I apologise for not searching thorough enough.
I was wondering if it's possible to somehow get HA to recognise the service parameters of a function decorated as `@service` like it does with the built-in services when you use them in an automation.

An example - if I create an automation using HA's automation UI and I want to turn on a light I geht these choices:

<img width="1056" alt="image" src="https://user-images.githubusercontent.com/304352/156187301-a107428a-aadf-4ad4-8fd8-fa505c69bf05.png">

HA knows the parameters the service accepts, shows a description for them and also offers specific input methods (check boxes, sliders or dropdowns) depending on the service and parameters it has.

While I can do a `Call Service` for a pyscript service it only has a very limited parameter selection (an entitiy dropdown):
<img width="1055" alt="image" src="https://user-images.githubusercontent.com/304352/156187569-c7d8a4a6-74f5-4539-b34d-7582ecc200f3.png">

Is there a way to somehow annotate my service to also get "fancy" options? 
Pyscript's own "Start Jupyter Notebook" also has the full UI choices so is it possible?

**Comment by HealsCodes** [comment, marked as answer]
Answering myself here in case anyone else comas across this!

pyscript supports passing a yaml description for a service as part of the functions docstring.
Parameter descriptions follow the same format described here for normal [configuration variables](https://www.home-assistant.io/integrations/script/#configuration-variables) in yaml scripts.

So this would do what I want:
```python
@service 
def light_rgb_turn_on(entity_id, transition=0):
  """yaml
  description: turn a HUE rgb light on with automatic parameter setting
  fields:
    entity_id:
      description: name of the entity
      example: light.bedroom
      required: true
      selector:
        entity:
    transition:
      description: duration in seconds to transition to the new state
      required: false
      selector:
        number:
          min: 0
          max: 60
          step: 1
          mode: slider
  """
  # rest of the functions body here
```

and produces this result:
<img width="677" alt="image" src="https://user-images.githubusercontent.com/304352/156196033-5c7cde8c-e5fc-4149-8610-1886e1025aa9.png">

**Comment by rct** [comment]
@HealsCodes Thanks for this Q and A. I was about to ask the same question because I couldn’t find it.

is this documented some where?

**Reply by HealsCodes** [reply]
I can't vouch that it was documented when I wrote that answer two years ago but it certainly is mentioned in the [current documentation](https://hacs-pyscript.readthedocs.io/en/stable/reference.html#service-service-name-supports-response-none) and also links to the HA developer docs to show the available syntax :)

**Reply by rct** [reply]
Thanks I couldn't find it in my first search of the docs.  On the second search from a desktop browser I was able to find the documentation for `@service` which is buried under 'Other Function Decoraters' after compile and executor.   After I found it, came back here to reply, but I see you beat me too it with your quick reply. Sorry if I wasted your time.

I was expecting defining services in PyScript to be a little more prominently documented.  I think implementing services in PyScript would be a desirable use case, particularly once the feature allowing services to return responses was implemented in 2023.7
